builder.go

```diff
@@ -175,25 +203,57 @@ func containsEntry(metRequires []string, entry BuildPlanEntry) bool {
 	return false
 }
 
-type processMap map[string]launch.Process
+type processMap struct {
+	typeToProcess map[string]launch.Process
+	defaultType   string
+}
+
+func newProcessMap() processMap {
+	return processMap{
+		typeToProcess: make(map[string]launch.Process),
+		defaultType:   "",
+	}
+}
 
-func (m processMap) add(l []launch.Process) {
-	for _, proc := range l {
-		m[proc.Type] = proc
+// This function adds the processes from listToAdd to processMap
+// it sets m.defaultType to the last default process
+// if a non-default process overrides a default process, it returns its type and unset m.defaultType
+func (m *processMap) add(listToAdd []launch.Process) string {
+	result := ""
+	for _, procToAdd := range listToAdd {
+		if procToAdd.Default {
+			m.defaultType = procToAdd.Type
+			result = ""
+		} else {
+			existingProc, ok := m.typeToProcess[procToAdd.Type]
+			if ok && existingProc.Type == m.defaultType { // existingProc.Default = true
+				// non-default process overrides a default process
+				m.defaultType = ""
+				result = procToAdd.Type
+			}
+		}
+		m.typeToProcess[procToAdd.Type] = procToAdd
 	}
+
+	return result
 }
 
+// list returns a sorted array of processes.
+// The array is sorted based on the process types.
+// The list is sorted for reproducibility.
 func (m processMap) list() []launch.Process {
 	var keys []string
-	for key := range m {
-		keys = append(keys, key)
+	for proc := range m.typeToProcess {
+		keys = append(keys, proc)
 	}
 	sort.Strings(keys)
-	procs := []launch.Process{}
+	result := []launch.Process{}
 	for _, key := range keys {
-		procs = append(procs, m[key])
+		processWithNoDefault := m.typeToProcess[key]
+		processWithNoDefault.Default = false // we set the default to false so it won't be part of metadata.toml
```

Comment:
Maybe we could add a `NoDefault()` function on `Process`, like we have with `bp.noAPI()` or `bp.noHomepage()`.

Reply:
I like this idea. We'll do it.